{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pysam\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.patches as patches\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "NM_PERCENTAGE_THRESHOLD = 0.05 # Contigs will not be displayed if NM_percentage >= NM_PERCENTAGE_THRESHOLD; parameter used to exclude mismapped contigs\n",
    "INSERTION_THRESHOLD = 10 # Insertions will not be displayed if length < INSERTION_THRESHOLD\n",
    "INSERTION_THRESHOLD_2 = 100 # Insertions will be highlighted if length >= INSERTION_THRESHOLD_2\n",
    "DELETION_THRESHOLD = 1 # Delections will not be displayed if length < DELETION_THRESHOLD\n",
    "ALIGN_CONTINGENCY = 10 # Blocks will not be displayed if length < ALIGN_CONTINGENCY\n",
    "\n",
    "RECTANGLE_HEIGHT = 0.2\n",
    "INTER_ASSEMBLY_PADDING = 1\n",
    "X_PADDING = 0.05"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Canvas:\n",
    "    def __init__(self):\n",
    "        self.assemblies = []\n",
    "        self.y = 0 - INTER_ASSEMBLY_PADDING\n",
    "        \n",
    "    def add_assembly(self, assembly):\n",
    "        assembly.y0 = self.y + INTER_ASSEMBLY_PADDING\n",
    "        self.assemblies.append(assembly)\n",
    "        self.y = max([assembly.y_max for assembly in self.assemblies])\n",
    "        \n",
    "    def draw(self, target_region = None): \n",
    "        # Determine x and y limits\n",
    "        x_min = None\n",
    "        x_max = None\n",
    "        y_max = 0\n",
    "        \n",
    "        for assembly in self.assemblies:\n",
    "            if x_min is None or x_min > assembly.left:\n",
    "                x_min = assembly.left\n",
    "            if x_max is None or x_max < assembly.right:\n",
    "                x_max = assembly.right\n",
    "            if y_max < assembly.y_max + assembly.y0:\n",
    "                y_max = assembly.y_max + assembly.y0\n",
    "        if target_region:\n",
    "            x_min = target_region[0]\n",
    "            x_max = target_region[1]\n",
    "        \n",
    "        # Create figure\n",
    "        plt.figure(figsize = (10, 0.5 * y_max))\n",
    "        \n",
    "        # Plot assemblies\n",
    "        for assembly in self.assemblies:\n",
    "            assembly.draw()\n",
    "        \n",
    "        # Set x and y limits\n",
    "        delta_x = X_PADDING * (x_max - x_min)\n",
    "        plt.xlim(x_min - delta_x, x_max + delta_x)\n",
    "        \n",
    "        plt.ylim(0, y_max + 1)\n",
    "        \n",
    "        # Label assemblies\n",
    "        for assembly in self.assemblies:\n",
    "            plt.text(x_max, assembly.y0 + 0.05, assembly.name, ha = 'right', va = 'bottom')\n",
    "        \n",
    "        # Hide ticks on y axis\n",
    "        plt.tick_params(axis='y', which='both', left=False, right=False, labelleft=False)\n",
    "        \n",
    "        # Misc\n",
    "        plt.xlabel(\"Reference postition\")\n",
    "        plt.tight_layout()\n",
    "                \n",
    "        \n",
    "    def __repr__(self):\n",
    "        output = \"\"\n",
    "        for assembly in self.assemblies:\n",
    "            output += assembly.__repr__()\n",
    "        return output\n",
    "    \n",
    "\n",
    "class Assembly:\n",
    "    def __init__(self, y0 = 0, name = \"\"):\n",
    "        self.rows = []\n",
    "        self.name = name\n",
    "        self.y0 = y0\n",
    "        self.y_max = 0\n",
    "        self.left = None\n",
    "        self.right = None\n",
    "        \n",
    "    def add_contig(self, contig):\n",
    "        if self.left is None or self.left > contig.left:\n",
    "            self.left = contig.left\n",
    "        if self.right is None or self.right < contig.right:\n",
    "            self.right = contig.right\n",
    "            \n",
    "        # Try to insert contig into one of the existing rows\n",
    "        for row in self.rows:\n",
    "            if row.check_overlap(contig) == False:\n",
    "                row.add_contig(contig)\n",
    "                return\n",
    "        \n",
    "        # Otherwise, create a new row and add the contig\n",
    "        y = self.y_max + 1\n",
    "        new_row = self.add_row(y)\n",
    "        self.y_max = y\n",
    "        new_row.add_contig(contig)\n",
    "    \n",
    "\n",
    "            \n",
    "    def add_row(self, y):\n",
    "        new_row = Row(y)\n",
    "        self.rows.append(new_row)\n",
    "        return new_row\n",
    "\n",
    "    def draw(self):\n",
    "        for row in self.rows:\n",
    "            row.draw(self.y0)\n",
    "            \n",
    "        # Draw a horizontal seperating line\n",
    "        plt.gca().axhline(y = self.y0, linewidth=1, color='whitesmoke')\n",
    "\n",
    "    def __repr__(self):\n",
    "        output = \"Assembly {name} (y0 = {y0})\\n\".format(name=self.name, y0=self.y0)\n",
    "        for row in self.rows:\n",
    "            output += row.__repr__()\n",
    "        return output\n",
    "        \n",
    "\n",
    "class Row:\n",
    "    def __init__(self, y):\n",
    "        self.y = y\n",
    "        self.contigs = []\n",
    "        \n",
    "    def check_overlap(self, contig, padding = 10):\n",
    "        overlapped = False\n",
    "        for existing_contig in self.contigs:\n",
    "            if contig.left >= existing_contig.left - padding and contig.left <= existing_contig.right + padding or \\\n",
    "            contig.right >= existing_contig.left - padding and contig.right <= existing_contig.right + padding:\n",
    "                overlapped = True\n",
    "        return overlapped\n",
    "    \n",
    "    def add_contig(self, contig):\n",
    "        self.contigs.append(contig)\n",
    "        \n",
    "    def draw(self, y0):\n",
    "        for contig in self.contigs:\n",
    "            contig.draw(self.y + y0)\n",
    "            \n",
    "    def __repr__(self):\n",
    "        output = \"Row {y}\\n\".format(y = self.y)\n",
    "        for contig in self.contigs:\n",
    "            output += contig.__repr__() + \"\\n\"\n",
    "        return output\n",
    "            \n",
    "            \n",
    "\n",
    "class Contig:\n",
    "    def __init__(self, contig_data):\n",
    "        self.blocks = []\n",
    "        self.left = None\n",
    "        self.right = None\n",
    "        self.width = None\n",
    "\n",
    "        # Add blocks\n",
    "        blocks_data = contig_data.get_blocks()\n",
    "        for block_data in blocks_data:\n",
    "            left, right = block_data[0], block_data[1]\n",
    "            block = Block(left, right)\n",
    "            self.add_block(block)\n",
    "            \n",
    "        # Add insertions and deletions\n",
    "        aligned_pairs = contig_data.get_aligned_pairs()\n",
    "        self.insertions = detect_insertions(aligned_pairs)\n",
    "        self.deletions = detect_deletions(aligned_pairs)\n",
    "        \n",
    "        # Compute statictics\n",
    "        self.NM = contig_data.get_tag('NM')\n",
    "        self.NM_percentage = self.NM / self. width\n",
    "        \n",
    "        # self.NM_percentage = sum([block.width + 1 for block in self.blocks]) / (self.width + 1)\n",
    "        \n",
    "        \n",
    "    def add_block(self, block):\n",
    "        self.blocks.append(block)\n",
    "        if self.left is None or self.left > block.left:\n",
    "            self.left = block.left\n",
    "        if self.right is None or self.right < block.left:\n",
    "            self.right = block.right\n",
    "        self.width = self.right - self.left\n",
    "            \n",
    "    def draw(self, y):\n",
    "        if self.NM_percentage >= NM_PERCENTAGE_THRESHOLD:\n",
    "            return\n",
    "        \n",
    "        # Draw a line across the contig\n",
    "        width = self.right - self.left\n",
    "        line_height = 0.05\n",
    "        ax = plt.gca()\n",
    "        global RECTANGLE_HEIGHT\n",
    "        ax.add_patch(patches.Rectangle((self.left, y + 0.5 * RECTANGLE_HEIGHT - 0.5 * line_height), \n",
    "                                       width, line_height, facecolor = \"firebrick\", edgecolor=None, zorder=0.8))\n",
    "        for block in self.blocks:\n",
    "            block.draw(y)\n",
    "        \n",
    "        if self.insertions:\n",
    "            for insertion in self.insertions:\n",
    "                insertion.draw(y)\n",
    "        \n",
    "        if self.deletions:\n",
    "            for deletion in self.deletions:\n",
    "                deletion.draw(y)\n",
    "        \n",
    "            \n",
    "    def __repr__(self):\n",
    "        return \"Contig {left}-{right} (NM% = {NM}; {n_insertion} insertions; {n_deletion} deletions; {n} blocks) \".format(\n",
    "            left = self.left, right = self.right, n = len(self.blocks),\n",
    "            NM = round(self.NM_percentage * 100, 2), n_insertion = len(self.insertions), n_deletion = len(self.deletions))\n",
    "\n",
    "class Block:\n",
    "    def __init__(self, left, right):\n",
    "        self.left = left\n",
    "        self.right = right\n",
    "        self.width = self.right - self.left\n",
    "        self.y = None\n",
    "        \n",
    "    def draw(self, y):\n",
    "        if self.right - self.left < ALIGN_CONTINGENCY:\n",
    "            return\n",
    "        else:\n",
    "            rectangle_color = \"steelblue\"\n",
    "        ax = plt.gca()\n",
    "        ax.add_patch(patches.Rectangle((self.left, y), self.width, RECTANGLE_HEIGHT, zorder=1, facecolor=rectangle_color, edgecolor=None))\n",
    "        \n",
    "class Insertion:\n",
    "    def __init__(self, position, length):\n",
    "        self.position = position\n",
    "        self.length = length\n",
    "    def draw(self, y):\n",
    "        global INSERTION_THRESHOLD\n",
    "        if self.length < INSERTION_THRESHOLD:\n",
    "            return\n",
    "        if not self.position or not y:\n",
    "            return\n",
    "        if self.length >= INSERTION_THRESHOLD_2:\n",
    "            marker_color = 'red'\n",
    "        else:\n",
    "            marker_color = 'purple'\n",
    "        plt.plot(self.position, y, marker = align_marker(\"^\", valign=\"top\"), color=marker_color, markersize=10, markeredgewidth=0)\n",
    "        \n",
    "    def __repr__(self):\n",
    "        return \"Insertion(position = {position}, length = {length})\".format(\n",
    "            position = self.position, length = self.length)\n",
    "    \n",
    "class Deletion():\n",
    "    def __init__(self, left, right):\n",
    "        self.left = left\n",
    "        self.right = right\n",
    "        self.width = self.right - self.left + 1\n",
    "    def __repr__(self):\n",
    "        output = \"Deletion(width = {width}) \".format(width = self.width)\n",
    "        return output\n",
    "    def draw(self, y):\n",
    "        if self.width < DELETION_THRESHOLD:\n",
    "            return\n",
    "        ax = plt.gca()\n",
    "        ax.add_patch(patches.Rectangle((self.left, y), self.width, RECTANGLE_HEIGHT, zorder=2, facecolor='red', edgecolor=None))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [],
   "source": [
    "def detect_insertions(aligned_pairs):\n",
    "    # How to distinguish between small deletions and poorly assemblied regions?\n",
    "    length = 0\n",
    "    position = 0\n",
    "    insertions = []\n",
    "    for query, reference in aligned_pairs:\n",
    "        if query and not reference:\n",
    "            if length == 0: # Start of a new insertion              \n",
    "                pass\n",
    "            else: # Continue current insertion\n",
    "                pass\n",
    "            length += 1\n",
    "            \n",
    "        else:\n",
    "            if length == 0: # Matched pair; record position\n",
    "                position = reference\n",
    "            else: # End current insertion\n",
    "                insertions.append(Insertion(position, length))\n",
    "                position = 0\n",
    "                length = 0\n",
    "    return insertions\n",
    "\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [],
   "source": [
    "def detect_deletions(aligned_pairs):\n",
    "    deletions = []\n",
    "    left = None\n",
    "    right = None\n",
    "    for query, reference in aligned_pairs:\n",
    "        if reference and not query: \n",
    "            if not left: # Start of a new deletion\n",
    "                left = reference\n",
    "                right = left + 1\n",
    "            else: # Continue current deletion\n",
    "                right = reference\n",
    "        else:\n",
    "            if not left: # Matched pair; record position\n",
    "                pass\n",
    "            else: # End current insertion\n",
    "                deletions.append(Deletion(left, right))\n",
    "                left = None\n",
    "                right = None\n",
    "    return deletions\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [],
   "source": [
    "from matplotlib import markers\n",
    "from matplotlib.path import Path\n",
    "\n",
    "def align_marker(marker, halign='center', valign='middle',):\n",
    "    \"\"\"\n",
    "    create markers with specified alignment.\n",
    "    From: https://stackoverflow.com/questions/26686722/align-matplotlib-scatter-marker-left-and-or-right\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "\n",
    "    marker : a valid marker specification.\n",
    "      See mpl.markers\n",
    "\n",
    "    halign : string, float {'left', 'center', 'right'}\n",
    "      Specifies the horizontal alignment of the marker. *float* values\n",
    "      specify the alignment in units of the markersize/2 (0 is 'center',\n",
    "      -1 is 'right', 1 is 'left').\n",
    "\n",
    "    valign : string, float {'top', 'middle', 'bottom'}\n",
    "      Specifies the vertical alignment of the marker. *float* values\n",
    "      specify the alignment in units of the markersize/2 (0 is 'middle',\n",
    "      -1 is 'top', 1 is 'bottom').\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "\n",
    "    marker_array : numpy.ndarray\n",
    "      A Nx2 array that specifies the marker path relative to the\n",
    "      plot target point at (0, 0).\n",
    "\n",
    "    Notes\n",
    "    -----\n",
    "    The mark_array can be passed directly to ax.plot and ax.scatter, e.g.::\n",
    "\n",
    "        ax.plot(1, 1, marker=align_marker('>', 'left'))\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    if isinstance(halign, (bytes, str)):\n",
    "        halign = {'right': -1.,\n",
    "                  'middle': 0.,\n",
    "                  'center': 0.,\n",
    "                  'left': 1.,\n",
    "                  }[halign]\n",
    "\n",
    "    if isinstance(valign, (bytes, str)):\n",
    "        valign = {'top': -1.,\n",
    "                  'middle': 0.,\n",
    "                  'center': 0.,\n",
    "                  'bottom': 1.,\n",
    "                  }[valign]\n",
    "\n",
    "    # Define the base marker\n",
    "    bm = markers.MarkerStyle(marker)\n",
    "\n",
    "    # Get the marker path and apply the marker transform to get the\n",
    "    # actual marker vertices (they should all be in a unit-square\n",
    "    # centered at (0, 0))\n",
    "    m_arr = bm.get_path().transformed(bm.get_transform()).vertices\n",
    "\n",
    "    # Shift the marker vertices for the specified alignment.\n",
    "    m_arr[:, 0] += halign / 2\n",
    "    m_arr[:, 1] += valign / 2\n",
    "\n",
    "    return Path(m_arr, bm.get_path().codes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_bam_file(bam_file, assembly_name = None):\n",
    "    if not assembly_name: \n",
    "        # Use bamfile name without extension as assembly name\n",
    "        base_name = os.path.basename(bam_file)\n",
    "        assembly_name = os.path.splitext(base_name)[0]\n",
    "    assembly = Assembly(name = assembly_name)\n",
    "    bam = pysam.AlignmentFile(bam_file, 'rb')\n",
    "    \n",
    "    contigs = []\n",
    "    for contig_data in bam.fetch():\n",
    "        contig = Contig(contig_data)    \n",
    "        contigs.append(contig)\n",
    "    \n",
    "    # Sort contigs by width to ensure longer contigs are plotted before shorter contigs\n",
    "    contigs.sort(key = lambda contig: contig.width, reverse=True)\n",
    "    for contig in contigs:\n",
    "        assembly.add_contig(contig)\n",
    "    return assembly"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAsgAAAEYCAYAAABBfQDEAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8QZhcZAAAgAElEQVR4nO3dfZQc1Xnn8d8jDVaDQRmBDAvC0gitAxgtL2YMeHmZ8ctBOCKYLBBQMMhg1gs2kPgc74bEK6Q1h7MxBtQHEgeCQQOYhBg2DC+CFawRsjFq2yMswNgEG5gWsjGygeoRxj16u/tHVQ93mu6efq9++X7OmTNdt6pu3XrqVs3Tt2u6zDknAAAAAKFpcTcAAAAAaCUkyAAAAICHBBkAAADwkCADAAAAHhJkAAAAwNNTaubs2bNdX19fk5oCAAAANM+GDRt+55z7QH55yQS5r69PIyMjjWsVAAAAEBMzSxcq5xYLAAAAwEOCDAAAAHhIkAEAAAAPCTIAAADgIUEGAAAAPCTIAAAAgIcEGQAAAPCQIAMAAAAeEmQAAADAQ4IMAAAAeEiQAQAAAA8JMgAAAOAhQQYAAAA8JMgAAACAhwQZAAAA8JAgAwAAAB4SZAAAAMBDggwAAAB4SJABAAAAT0/cDQAAAM216KrV7ylbs2xxDC0BWhMjyAAAAICHBBkAAADwkCADAAAAHhJkAAAAwEOCDAAAAHj4FgsAALoM31gBlMYIMgAAAOAhQQYAAAA8JMgAAACAhwQZAAAA8JAgAwAAAB4SZAAAAMBDggwAAAB4SJABAAAADwkyAAAA4CFBBgAAADwkyAAAAICHBBkAAADwkCADAAAAHhJkAAAAwEOCDAAAAHhIkAEAAAAPCTIAAADgIUEGAAAAPCTIAAAAgKcn7gYAAIB4LbpqdcHyNcsWN7klQGtgBBkAAADwkCADAAAAHhJkAAAAwEOCDAAAAHhIkAEAAAAP32IBAECXW7NssYLRQGObxyRJMw+cGXOLgHiRIAMA0OWyQVa3nXCbtv56q2TSnvvtqWk903TJs5co0ZuIu3lA03GLBQAAXS6VTGnrr7ZKTtIu6e3X3tbYq2NKJVNxNw2IBQkyAABdLBtk9dT1TxWcl1qZUjbINrlFQPxIkAEA6GKpZErbt24vOG98bJxRZHQlEmQAALpYojehfQ7eRz2798imm2SS9Zh232d3LVi0gHuQ0ZXMOVd0Zn9/vxsZGWlicwAAAIDmMLMNzrn+/HJGkAEAAAAPCTIAAADgIUEGAAAAPCTIAAAAgIcEGQAAAPCQIAMAAAAeEmQAAADAQ4IMAAAAeEiQAQAAAA8JMgAAAOAhQQYAAAA8JMgAAACAhwQZAAAA8JAgAwAAAJ6euBsANNKiq1YXLF+zbHHd665Hnd2u1WLaau0BGqHYdTKnGf2+VBs47xAHRpABAAAADwkyAAAA4CFBBgAAADwkyAAAAIDHnHNFZ/b397uRkZEmNgdovGA0UG9fb13r2fTkJs09YW7NdWJyLOt1rOrVHqDTBaOBfnHHU9rvEwvp9+gKZrbBOdefX84IMrpKNshqaHBI2SBbt3qCdKChwSEF6aBOrexefizrdazq1R6g02WDrIaOuVEPL/+Rhgbo9+huJMjoKqlkSpl0Rqlkqm71DC8dltvpNLx0uE6t7F5+LOt1rOrVHqDTpZIpZX/7e0kmt4t+j+5GgoyukQ2yE8lWKpmqemTSr+ep659Sel1akpRel2bEpQZBOpgUy/XXr5dU27GqZ3s4tuhk2SCrH3/jcY0rMVFGv0c3I0FG10glUxrPjEuSxjPjVY9M+vVs37p90jxGXKqXH7ttW7dJqu1Y1bM9HFt0slQypZ3vbJNkk8rp9+hWPEkPXSPRm9DA8oFJ07XW89O7f6pdO3Zp5oEzw3mzqqsTYezmDcyTJI1tHtO0nmlaeM7CcF6Vx6pe7clNA50q0ZvQbu+bpu3bdspFSfK0aUa/R9fiWywAAADQlfgWCwAAAKAMJMgAAACAhwQZAAAA8JAgAwAAAB4SZAAAAMBDggwAAAB4SJABAAAADwkyAAAA4CFBBgAAADwkyAAAAICHBBkAAADwkCADAAAAHhJkAAAAwEOCDAAAAHh64m4AUA+Lrlpd1nJrli2uuu78dfO3WU3dnabYcSgnNoXW9dfz51cS66nqrfd6KP98bHft1h9qPS713t9y2tNuMUbnYAQZAAAA8JAgAwAAAB4SZAAAAMBDggwAAAB4zDlXdGZ/f78bGRlpYnMAAO0uGA3U29c7ZVm9122UuLcfh01PbtLcE+Y2Z2Ojo1Jf33uKuzHuaD4z2+Cc688vZwQZAFA32SCrocEhZYNsybJ6r9socW8/DkE60NDgkIJ00ISNBdLgYPjb041xR2shQQYA1E0qmVImnVEqmSpZVu91GyXu7cdheOmw3E6n4aXDjd9YMiml0+FvTzfGHa2FBBkAUBfZIDuR0KSSKWWDbMGyeq/bKHFvPw5BOlB6XVqSlF6XbuwochC8mxgnkxOjyN0Yd7QeEmQAQF2kkimNZ8YlSeOZcaWSqYJl9V63UeLefhzyR40bOoqcTEqZTPg6k5lIlrsx7mg9PEkPAFAXid6EBpYPTJqWVLCsnus2SrE2dbLErITmDcybNN0wvb3S8uWTp9WdcUfr4VssAAAA0JX4FgsAAACgDCTIAAAAgIcEGQAAAPCQIAMAAAAeEmQAAADAQ4IMAAAAeEiQAQAAAA8JMgAAAOAhQQYAAAA8JMgAAACAhwQZAAAA8JAgAwAAAB4SZAAAAMBDggwAAAB4euJuQL5FV62Ouwloc2uWLS5Y3ui+VWy7zWxDqyknJr44jlG3HZNmqfTYt6N27jtrli2uuv21rFuqTql0TEv1qXY+FnhXK103GEEGAAAAPCTIAACgbA8uWBB3E4CGI0EGAABl+9OXXoq7CUDDkSADAAAAHnPOFZ3Z39/vRkZGmticBhgdlfr6FIwG6u3rnXLxcpcrZtOTmzT3hLlVr9/qao1PqfWD0UCSaqq/nmrdV9Ruqv7SlccnuqY1Q6Ux7rZj4u9vsX0vtzy/Lqnya2E18W+JY9bEPt2WaojPVMc3N78l+kFMzGyDc64/v7yzR5CDQBocVDb9uoYGh5QNsiUXzwbZspYrurl0oKHBIQXpoKr1W12t8Sm1fjbIatVJq7TqpFVV119Pte4rajdVf+nK4xNd0xQ0/hpTaYy77Zj4+1ts38stz6+rmmthNfFviWPWxD7dlmqIz1THNzc/l7t0y7lbrs5OkJNJKZ1Wauk/KpPOKJVMlVw8lUyVtVwxw0uH5XY6DS8drmr9VldrfEqtn0qmNPbqmMZeHau6/nqqdV9Ru6n6S1cen+iapmSy4ZuqNMbddkz8/S227+WW59dVzbWwmvi3xDFrYp9uSzXEZ6rjm5s/vHQ4/n7Qgjo3QQ4CKZlUVgml1o1LCjtDqXdS/sWq0ndSQTpQel1akpRel+64UeRa41Nq/WyQ1fqV6yemUysrr7+eat1X1G6q/tKVxye6pkkKfzdwxK3SGHfbMfH3d/3K9UqtfO++F4tJfnmQDibVtf76yq+F1cS/JY5ZE/t0W6ohPlMdX39+LnfphnO3Ep2bICeTUiajlI7TuBKSpPHMeMl3UuOZ8SmXKyZ/1LjTRpFrjU+p9VPJlLaNbZuYHh+rvP56qnVfUbup+ktXHp/omiYp/N3AEbdKY9xtx8Tf321j2zQ+9t59LxaT/PLhpcOT6tq2tfJrYTXxb4lj1sQ+3ZZqiM9Ux9efn9MN524lWu5JenXT2ystX65EShrIOimRkI47ToneRMHFE70JDSwfmDRdicSshOYNzJs03Ulqjk+J9RO9CS1YtEA7sjskST2Jnorrr6da9xW1m6q/dOXxia5pk6YbpNIYd9sx8fd3c2qzJOnA4w6cmJe/TKny3zzzm0l1BaOB9vwPe0oq/1pYTfxb4pg1sU+3pRriM9Xxzc3fnNqsHdkd6kn06MDjDuz4c7cSnf8tFgAAAEAB3fktFgAAAECFSJABAAAADwkyAAAA4CFBBgAAADwkyAAAAICHBBkAAADwkCADAAAAHhJkAAAAwEOCDAAAAHhIkAEAAAAPCTIAAADgIUEGAABAUXfccYcOP/xwHXHEETrvvPP04IMP6thjj9VRRx2lT33qU3r99dclSStWrNCFF16owcFBHXTQQbrhhhskSaOjo1q4cOFEfddee61WrFgRx66UrSfuBgAAAKA1Pf/887r66qv1gx/8QLNnz9abb74pM1MqlZKZ6Vvf+pauueYaXXfddZKkF154QWvXrtXWrVt18MEH65JLLol5D6pDggwAAICCHn/8cZ155pmaPXu2JGnvvffWc889p7PPPluvvfaatm3bpvnz508sv3jxYs2YMUMzZszQvvvuOzG63G64xQIAAAAFOedkZpPKLrvsMl166aV67rnndPPNNyubzU7MmzFjxsTr6dOna8eOHerp6dGuXbsmyv3lWxUJMgAAAAr65Cc/qe985zt64403JElvvvmmMpmM5syZI0m6/fbbp6xjv/3205YtW/TGG29ofHxcDz30UEPbXA/cYgEAAICCDjvsMH31q1/VwMCApk+frqOOOkorVqzQWWedpTlz5ui4447TK6+8UrKO3XbbTVdeeaWOPfZYzZ8/X4ccckiTWl89c84Vndnf3+9GRkaa2BwAAACgOcxsg3OuP7+cWywAAAAAT8lbLHbt2qV33nmnWW0BAAAAYlcyQZ42bZr22GOPZrUFAAAAiB23WAAAAAAeEmQAAADAQ4IMAAAAeEiQAQAAAA8JMgAAAOAhQQYAAAA8JMgAAACAhwQZAAAA8JAgAwAAAB4SZAAAAMBDggwAAAB4SJABAAAADwkyAAAA4CFBBgAAADwkyAAAAICHBBkAAADwkCADAAAAHhJkAAAAwEOCDAAAAHh64m4A2tOiq1bXXMeaK0/Voq89VIfWFKl/2eKG1S29NwaN3l5XMJvUJ/yYltPnCvWpOI5LPc6PctHvAKD+GEEGAAAAPCTIAAAAgIcEGQAAAPCQIAMAAAAec84Vndnf3+9GRkaa2Bx0umA0UG9fb9F5korObzWl9gX1UU2MO/64jI5KfX1xtwIAOoKZbXDO9eeXM4KMpskGWQ0NDikbZAvOW3XSKq06aVXB+a2m1L6gPqqJcccflyCQBgfD3wCAhiFBRtOkkill0hmlkqmC88ZeHdPYq2MF57eaUvuC+qgmxh1/XJJJKZ0OfwMAGoYEGU2RDbITSUsqmZo0wpcNslq/cv3EdGplqqVHAEvtC+qjmhh3/HEJgncT42SSUWQAaCASZDRFKpnSeGZckjSeGZ80wpdKprRtbNvE9PjYeEuPAJbaF9RHNTHu+OOSTEqZTPg6k2EUGQAaiCfpoSkSvQkNLB+YNO2/XrBogXZkd0iSehI9k+a3mlL7gvqoJsYdf1x6e6XlyydPAwAagm+xAAAAQFfiWywAAACAMpAgAwAAAB4SZAAAAMBDggwAAAB4SJABAAAADwkyAAAA4CFBBgAAADwkyAAAAICHBBkAAADwkCADAAAAHhJkAAAAwEOCDAAAAHhIkAEAAAAPCTIAAADg6Ym7AfkWXbU6tm2vWbY41u23kzXLFsfdhJpUc5zXXHmqFn3toaq3metftdbTyUrFJo4+l99Pcm0ot/80o82tfM0qtP+t3N5iyjmO7bhfOfn7N9W+1PtvZaXbR3Xa5W9Pq+QXjCADAAAAHhJkAABQtgcXLIi7CUDDkSADAAAAHhJkAABQtj996aW4mwA0HAkyAAAA4DHnXNGZ/f39bmRkpInNaYDRUamvr2mb2/TkJs09YW7TttdswWig3r7euJuBNhaMBpJUVj+iv8Wr0vg36ni1Qz8o1sZKy3PzpPLOkUbY9OQmzTxwZtO2/+I/PKo//tLJTdlWW6ohjyn33GmHc6xRzGyDc64/v7yzR5CDQBocDH83Y3PpQEODQwrSzdles2WDrIYGh5QNsnE3BW0qG2S16qRVWnXSqin7Ef0tXpXGv1HHqx36QbE2Vlqem1fuOdIIQTrQqoFVuu2E25qy/S1Pvqh9L/1zbXnyxYZvqy3VkMeUe+60wzkWh85OkJNJKZ0OfzfB8NJhuZ1Ow0uHm7K9ZkslU8qkM0olU3E3BW0qlUxp7NUxjb06NmU/or/Fq9L4N+p4tUM/KNbGSstz88o9RxpheOmwtEva+qutTdn+y6derl5l9PKplzd8W22phjym3HOnHc6xOHRughwE73aoZLLho8hBOlB6XVqSlF6X7rhR5GyQnTh5UskU7zRRsWyQ1fqV6yemUyuL9yP6W7wqjX+jjlc79INibay0PDev3HOkEfy/Y5K0/vr1Dd3+lidf1JGZJyRJR2aeYBQ5Xw15TLnnTjucY3Hp3AQ5mZQymfB1JtPwUeT8UeNOG0VOJVMaz4xLksYz47zTRMVSyZS2jW2bmB4fK96P6G/xqjT+jTpe7dAPirWx0vLcvHLPkUbI/7u1beu2hm7/5VMvV0JhLBIaZxQ5Xw15TLnnTjucY3FpuUdN101vr7R8+eTpBkrMSmjewLxJ050k0ZvQwPKBSdNAJRK9CS1YtEA7sjskST2JnqL9iP4Wr0rj36jj1Q79oFgbKy3PvS73HGmExKyEZi2YpV07djVl+zt231MpLZo0DU8NeUy55047nGNx6fxvsQAAAAAK6M5vsQAAAAAqRIIMAAAAeEiQAQAAAA8JMgAAAOAhQQYAAAA8JMgAAACAhwQZAAAA8JAgAwAAAB4SZAAAAFTshhtu0KGHHqo5c+bo0ksvjbs5ddW5j5oGAABAw3zzm9/UI488onXr1qnTnrzMCDIAAAAqcvHFF+vll1/WaaedprfeekuStHXrVs2fP1/bt2+XJI2Njamvr0/bt2/XSy+9pFNOOUVHH320TjzxRL3wwgtxNn9KJMgAAACoyE033aQDDjhAa9eu1axZsyRJe+21lwYHB7V69WpJ0t13360zzjhDu+22m77whS/oxhtv1IYNG3Tttdfqi1/8YpzNnxK3WAAAAKAuLrroIl1zzTU6/fTTtWrVKt1yyy16++239dRTT+mss86aWG58fDzGVk6NBBkAAAB1cfzxx2t0dFTr1q3Tzp07tXDhQo2Njam3t1cbN26Mu3ll4xYLAAAA1M3555+vJUuW6IILLpAkzZw5U/Pnz9c999wjSXLO6ZlnnomziVMiQQYAAEDdnHvuuXrrrbe0ZMmSibK77rpLt956q4444ggddthhuv/++2Ns4dTMOVd0Zn9/v+u0r+0AAABA49x77726//77deedd8bdlCmZ2QbnXH9+OfcgAwAAoC4uu+wyPfLII3r44YfjbkpNSJABAABQFzfeeGPcTaiLkrdYmNlvJaWb15yWNlvS7+JuRBsjftUjdtUjdrUhftUjdtUjdrUhfpWZ55z7QH5hyQQZ7zKzkUL3qKA8xK96xK56xK42xK96xK56xK42xK8++BYLAAAAwEOCDAAAAHhIkMv3T3E3oM0Rv+oRu+oRu9oQv+oRu+oRu9oQvzrgHmQAAADAwwgyAAAA4CFBBgAAADxdmSCb2QfNbK2Z/dzMnjezvyywzB+Z2YNm9ky0zAXevKVm9ovoZ6lXfrSZPWdmvzSzG8zMmrVPzVJL7MzsSDNbH5U9a2Zne+sMmdkrZrYx+jmymfvVLHXoezu9GD3glc83sx9GffJfzex9zdqnZqmx733ci9tGM8ua2enRvI7ve2XGbpaZ3Redmz8ys4XevFPM7N+ja9sVXjn9TsVjV2pdM1thZr/y+t2fNHO/mqUOfW80+ru60cxGvPK9zeyxqO89ZmazmrVPzVJj3zs475o3ZmZ/Fc3rir5XM+dc1/1I2l/SR6LXe0l6UdKH85b5W0lfj15/QNKbkt4naW9JL0e/Z0WvZ0XL/UjSxySZpEckfTrufW2x2P2xpA9F5QdIek1SbzQ9JOnMuPevleMXTb9dpN7vSDonen2TpEvi3tdWi523zN5R+R7RdMf3vTJj9w1Jy6PXh0j6bvR6uqSXJB0UncfP5Nal300Zu6LrSloh6Stx718rxy+aHpU0u0C910i6Inp9Re6876SfWmPnLTNd0m8UPhCja/perT9dOYLsnHvNOfd09HqrpJ9LmpO/mKS9zMwk7anwD+oOSYskPeace9M595akxySdYmb7S5rpnFvvwh54h6TTm7NHzVNL7JxzLzrnfhGt+2tJWxQmMV2jxr5XULTcJyTdGxXdLvpeqdidKekR59w7DW5yyygzdh+W9N1omRck9ZnZfpKOkfRL59zLzrltku6W9Bn63SQFY1fmuh2txr5XymcU9jmJvjdV7D4p6SXnHE9GrkBXJsg+M+uTdJSkH+bN+ntJh0r6taTnJP2lc26Xws75qrfc5qhsTvQ6v7xjVRE7f91jFI5GveQVXx19TLTSzGY0qt2tosr4JcxsxMxSuVsEJO0jKXDO5RJB+l6JvifpHEn/klfWNX2vROyekfRfomWOkTRP0oEqfs2j372rWOymWvfSqN/d1om3COSrMn5O0qNmtsHMvuCts59z7jUpTCQl7du4lsevlr6nwte8rup71ejqBNnM9pT0fyT9lXNuLG/2IkkbFd4KcKSkvzezmQpvn8jnSpR3pCpjl1t3f0l3SrrAS17+RuHHQx9V+BH4Xzd2D+JVQ/zmuvARon8hKWlmC0Tf85XT9/6TpDXeOl3T96aI3d9JmmVmGyVdJuknCkffueap6tiVWvcfJS1Q2E9fk3RdY/cgXjXE73jn3EckfVrSl8zspGa1uVXU2PfeJ+k0Sfd463RV36tW1ybIZrabwg53l3Pu3woscoGkf3OhX0p6ReEf0c2SPugtd6DC0arNmvyuLVfecWqInaJkZbWk/+mcS+VWiD5Kcs65cUmrFH6s25FqiV90a4qccy9LekLhiMLvJPWaWU+0Pn2vQOwify7pPufc9lxBt/S9qWLnnBtzzl3gnDtS0vkKb396RcWvefS7SInYFV3XOfe6c25nNEhwizq030m1xc+75m2RdJ/ejdPr0Rve3BvfLQ3fkRjUErvIpyU97Zx73Vuna/peLboyQY7unbtV0s+dc9cXWWyTwvt2FN3Pc7DCf8hbI+nk6D9HZ0k6WdKa6COerWZ2XFT/+ZLub/CuNF0tsYveyd4n6Q7nnP9uVt6FzhTeS/bTxuxBvGqM36zcx/9mNlvS8ZJ+Ft3zvlbhvbWStFT0vfzzNmeJ8j5q7Ia+V07szKzX3v0WioskfS8arfqxpA9Z+I0V71P4ce0D9LtJyxSMXal1c/0u8mfqwH4n1Ry/95vZXtEy71f49zYXpwcU9jmJvlfovM0pes2LdGzfq1VXPknPzE6Q9H2F9yjmPuL/W0lzJck5d5OZHaDwv9v3V/hR4t85574drX9htLwkXe2cWxWV90fr7K7wWywucx0W4FpiZ2afVThC97xX5eeccxvN7HGF73xN4UfkFzvn3m7CLjVVjfH7z5JujtabJinpnLs1qvcghf88tbfCj9g+G42Idow6nLd9kn4g6YP+fcnd0PfKjN3HFP5z8U5JP5P0eRf+I7Is/BqopML/hr/NOXd1VE6/KxG7Yus65x42szsVfsTtFH5Tw3/L3VPbSWqM30EKB1UkqUfSP3t9bx+F36IyV+Eb47Occ282Z6+aow7n7R4K/3/gIOdcxqu3K/perboyQQYAAACK6cpbLAAAAIBiSJABAAAADwkyAAAA4CFBBgAAADwkyAAAAGgr0VMAt5jZlF9TZ+FTUjdGPy+aWTDVOiTIADqKme2MLoI/NbMHzay3jHUuN7Ofm9ldzWhjXMxsMPq6wNz0xWZ2fvT6c9HX5OXmfcvMPhxHOwGgDEOSTilnQefcl51zR0YPVLlRUqEHTU1Cggyg0/whuhAulPSmpC+Vsc4XJf2Jc+7ccjbgPT2u3QxKmkiQnXM3OefuiCY/p/AR3bl5FznnftbU1gFAmZxz31N4jZ9gZgvM7P+a2QYz+76ZHVJg1fc8PKUQEmQAnWy9pDm5CTP772b2YzN71sz+V1R2k6SDJD1gZl+Ont51W7TcT8zsM9FynzOze8zsQUmPlqivLxqNvsXMnjezR81s92jefzSz/2dmz5jZ02a2oFg9+czsbTO7Llrvu2b2gaj8SDNLReveZ+ETPnOj4j+Lyu+OHpRysaQvRyPsJ5rZCjP7ipmdKalf0l3RvN3N7Ino4UcysyVm9lw0Kv/1vDZdHe1PysKnFwJAXP5J4UPajpb0FUnf9Gea2TxJ8yU9PlVFJMgAOpKZTVf42OkHoumTJX1I0jEKnyJ1tJmd5Jy7WNKvJX3cObdS0lclPe6c+6ikj0v6hoWPuZWkj0la6pz7RLH6ouU+JOkfnHOHSQoknRGV3xWVH6FwJPe1KerxvV/S0865j0haJ2l5VH6HpL92zh2u8IlbufIrJB0VlV/snBuVdJOkldEI+/dzFTvn7pU0IuncaN4fvDgeIOnrkj4Rte+jZna616ZUtD/fk/RfCx8NAGgsM9tT4XX1HjPbqPDJs/vnLXaOpHudczunqo8EGUCn2T26OL6h8BHIj0XlJ0c/P5H0tKRDFCam+U6WdEVUxxOSEooe7SrpMe9xtqXqe8U5tzF6vUFSn5ntJWmOc+4+SXLOZZ1z71TQrl2S/jV6/W1JJ5jZH0nqdc6ti8pvl5RLrp9VOCL8WUk7CoeqLB+V9IRz7rfOuR0Kk/zcNrZJesjfzxq2AwC1mCYpyN1rHP0cmrfMOSrj9gopfLY5AHSSPzjnjoySx4cU3oN8gyST9L+dczdPsb5JOsM59++TCs2OlfT7vOXeU190K8O4V7RT0u7R8sW2V0678rkp5i9WmMieJmmZmR1WYf05xdotSdudc7l27BR/UwDExDk3ZmavmNlZzrl7zMwkHe6ce0aSzOxgSbMU3no3JUaQAXQk51xG0uWSvmJmu0laI+nC6GM4mdkcM9u3wKprJF0WXVxlZkcV2US59eXaMyZpc+72BDObYWZ7VFDPNElnRq//QtKT0T6+ZWYnRuXnSVpnZtMkfdA5t1bS/5DUK2lPSVsl7VWkicXm/VDSgJnNjm5bWaLwFg8AiE43O1QAAAEISURBVI2Z/YvCZPdgM9tsZp+XdK6kz5vZM5Kel/QZb5Ulku723tSXxLt9AB3LOfeT6EJ5jnPuTjM7VNL6KPd9W9JnJW3JW+0qSUlJz0ZJ8qikUwvU/WiR+krd23aepJvN7GuStks6q0Q9+e36vaTDzGyDpIyks6PypZJuipLtlyVdIGm6pG9Ho+im8L7jIPoHw3ujfzy8LK/+oaiePyi81zq3n6+Z2d9IWhvV9bBz7v4S+wgADeecW1JkVsGvfnPOraikfiszkQYAxMjM3nbO7Rl3OwCgG3CLBQAAAOBhBBkAAADwMIIMAAAAeEiQAQAAAA8JMgAAAOAhQQYAAAA8JMgAAACA5/8DIcFBUCFaeBQAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 720x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "assembly = read_bam_file(\"data/IL27/flye.bam\")\n",
    "assembly2 = read_bam_file(\"data/IL27/canu.bam\")\n",
    "canvas = Canvas()\n",
    "canvas.add_assembly(assembly)\n",
    "canvas.add_assembly(assembly2)\n",
    "canvas.draw(target_region = (27888679, 29888679))\n",
    "\n",
    "plt.savefig(\"1.svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Assembly flye (y0 = 0)\n",
       "Row 1\n",
       "Contig 29498739-29901058 (NM% = 1.17; 45 insertions; 359 deletions; 404 blocks) \n",
       "Contig 27979949-28213983 (NM% = 0.71; 27 insertions; 122 deletions; 149 blocks) \n",
       "Contig 28458068-28656985 (NM% = 0.99; 28 insertions; 158 deletions; 186 blocks) \n",
       "Contig 29236568-29385329 (NM% = 1.03; 42 insertions; 79 deletions; 121 blocks) \n",
       "Contig 29067393-29209012 (NM% = 0.72; 37 insertions; 69 deletions; 106 blocks) \n",
       "Contig 28927451-29051025 (NM% = 1.14; 7 insertions; 108 deletions; 115 blocks) \n",
       "Contig 27877925-27976915 (NM% = 1.86; 11 insertions; 48 deletions; 59 blocks) \n",
       "Contig 29396994-29486875 (NM% = 1.98; 54 insertions; 95 deletions; 149 blocks) \n",
       "Contig 28658279-28709361 (NM% = 2.16; 15 insertions; 58 deletions; 73 blocks) \n",
       "Contig 28405943-28456774 (NM% = 1.38; 12 insertions; 57 deletions; 69 blocks) \n",
       "Contig 28363508-28405862 (NM% = 0.97; 8 insertions; 36 deletions; 44 blocks) \n",
       "Contig 28817853-28848081 (NM% = 0.71; 0 insertions; 17 deletions; 17 blocks) \n",
       "Contig 28799912-28811312 (NM% = 1.65; 8 insertions; 12 deletions; 20 blocks) \n",
       "Row 2\n",
       "Contig 28212831-28340121 (NM% = 1.05; 30 insertions; 99 deletions; 129 blocks) \n",
       "Contig 28843756-28928242 (NM% = 1.83; 7 insertions; 88 deletions; 95 blocks) \n",
       "Contig 28709344-28750839 (NM% = 1.43; 11 insertions; 37 deletions; 48 blocks) \n",
       "Contig 28685719-28698452 (NM% = 2.36; 10 insertions; 22 deletions; 32 blocks) \n",
       "Contig 28416644-28429291 (NM% = 2.22; 10 insertions; 21 deletions; 31 blocks) \n",
       "Assembly canu (y0 = 3)\n",
       "Row 1\n",
       "Contig 27879211-28336588 (NM% = 0.52; 88 insertions; 73 deletions; 161 blocks) \n",
       "Contig 29539711-29898423 (NM% = 0.19; 68 insertions; 63 deletions; 131 blocks) \n",
       "Contig 28778132-29047373 (NM% = 0.41; 43 insertions; 70 deletions; 113 blocks) \n",
       "Contig 28459923-28648560 (NM% = 0.13; 31 insertions; 20 deletions; 51 blocks) \n",
       "Contig 29227341-29377006 (NM% = 0.53; 39 insertions; 28 deletions; 67 blocks) \n",
       "Contig 29405489-29489409 (NM% = 1.05; 49 insertions; 41 deletions; 90 blocks) \n",
       "Contig 28685661-28760926 (NM% = 0.86; 29 insertions; 45 deletions; 74 blocks) \n",
       "Contig 28354302-28425040 (NM% = 0.2; 19 insertions; 23 deletions; 42 blocks) \n",
       "Contig 21387073-21421384 (NM% = 10.68; 46 insertions; 43 deletions; 89 blocks) \n",
       "Contig 21831573-21858560 (NM% = 4.12; 17 insertions; 37 deletions; 54 blocks) \n",
       "Contig 22512619-22539507 (NM% = 4.23; 23 insertions; 39 deletions; 62 blocks) \n",
       "Contig 29378118-29404983 (NM% = 2.95; 12 insertions; 25 deletions; 37 blocks) \n",
       "Contig 30219714-30246275 (NM% = 3.78; 20 insertions; 31 deletions; 51 blocks) \n",
       "Contig 29495614-29518258 (NM% = 1.69; 9 insertions; 13 deletions; 22 blocks) \n",
       "Contig 29226832-29227249 (NM% = 8.15; 2 insertions; 2 deletions; 4 blocks) \n",
       "Contig 12053322-12053518 (NM% = 4.59; 1 insertions; 1 deletions; 2 blocks) \n",
       "Row 2\n",
       "Contig 29044302-29222519 (NM% = 0.47; 30 insertions; 48 deletions; 78 blocks) \n",
       "Contig 28327643-28359113 (NM% = 0.44; 8 insertions; 19 deletions; 27 blocks) \n",
       "Contig 21387136-21414743 (NM% = 4.05; 35 insertions; 54 deletions; 89 blocks) \n",
       "Contig 29480374-29506771 (NM% = 3.19; 22 insertions; 28 deletions; 50 blocks) \n",
       "Contig 29366268-29391722 (NM% = 2.81; 22 insertions; 39 deletions; 61 blocks) \n",
       "Contig 22507923-22531549 (NM% = 0.84; 11 insertions; 27 deletions; 38 blocks) \n",
       "Contig 30227873-30251447 (NM% = 0.26; 9 insertions; 14 deletions; 23 blocks) \n",
       "Contig 21839820-21863258 (NM% = 1.72; 11 insertions; 22 deletions; 33 blocks) \n",
       "Contig 21415092-21437802 (NM% = 2.22; 8 insertions; 26 deletions; 34 blocks) \n",
       "Contig 29394083-29416463 (NM% = 1.78; 20 insertions; 22 deletions; 42 blocks) \n",
       "Contig 28635690-28654654 (NM% = 0.18; 6 insertions; 7 deletions; 13 blocks) \n",
       "Contig 28751784-28768777 (NM% = 1.65; 23 insertions; 18 deletions; 41 blocks) \n",
       "Contig 28683316-28698202 (NM% = 0.56; 9 insertions; 12 deletions; 21 blocks) \n",
       "Contig 28416819-28431693 (NM% = 0.11; 5 insertions; 7 deletions; 12 blocks) \n",
       "Contig 28797509-28811320 (NM% = 0.59; 8 insertions; 11 deletions; 19 blocks) \n",
       "Contig 28709344-28717466 (NM% = 0.2; 2 insertions; 5 deletions; 7 blocks) \n",
       "Contig 28397838-28405363 (NM% = 0.24; 2 insertions; 5 deletions; 7 blocks) \n",
       "Row 3\n",
       "Contig 21398971-21426280 (NM% = 4.94; 27 insertions; 36 deletions; 63 blocks) \n",
       "Contig 29487903-29511465 (NM% = 0.08; 5 insertions; 10 deletions; 15 blocks) \n",
       "Contig 29386365-29409688 (NM% = 3.17; 25 insertions; 33 deletions; 58 blocks) \n",
       "Contig 30235586-30258369 (NM% = 1.7; 10 insertions; 13 deletions; 23 blocks) \n",
       "Contig 21847548-21870070 (NM% = 1.43; 9 insertions; 24 deletions; 33 blocks) \n",
       "Contig 22500971-22523455 (NM% = 2.29; 9 insertions; 26 deletions; 35 blocks) \n",
       "Contig 28346561-28363510 (NM% = 1.73; 26 insertions; 12 deletions; 38 blocks) \n",
       "Contig 21828384-21844491 (NM% = 5.17; 22 insertions; 21 deletions; 43 blocks) \n",
       "Contig 28685718-28698452 (NM% = 1.45; 15 insertions; 13 deletions; 28 blocks) \n",
       "Contig 28416644-28429292 (NM% = 1.34; 13 insertions; 11 deletions; 24 blocks) \n",
       "Contig 28799911-28811320 (NM% = 0.71; 8 insertions; 8 deletions; 16 blocks) \n",
       "Row 4\n",
       "Contig 21407369-21430980 (NM% = 0.74; 10 insertions; 25 deletions; 35 blocks) \n",
       "Contig 29387648-29406365 (NM% = 3.74; 44 insertions; 21 deletions; 65 blocks) \n",
       "Row 5\n",
       "Contig 21396279-21412036 (NM% = 3.78; 31 insertions; 21 deletions; 52 blocks) \n",
       "Contig 29375403-29390961 (NM% = 1.12; 15 insertions; 10 deletions; 25 blocks) "
      ]
     },
     "execution_count": 107,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "canvas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on AlignedSegment object:\n",
      "\n",
      "class AlignedSegment(builtins.object)\n",
      " |  AlignedSegment(AlignmentHeader header=None)\n",
      " |  Class representing an aligned segment.\n",
      " |  \n",
      " |      This class stores a handle to the samtools C-structure representing\n",
      " |      an aligned read. Member read access is forwarded to the C-structure\n",
      " |      and converted into python objects. This implementation should be fast,\n",
      " |      as only the data needed is converted.\n",
      " |  \n",
      " |      For write access, the C-structure is updated in-place. This is\n",
      " |      not the most efficient way to build BAM entries, as the variable\n",
      " |      length data is concatenated and thus needs to be resized if\n",
      " |      a field is updated. Furthermore, the BAM entry might be\n",
      " |      in an inconsistent state.\n",
      " |  \n",
      " |      One issue to look out for is that the sequence should always\n",
      " |      be set *before* the quality scores. Setting the sequence will\n",
      " |      also erase any quality scores that were set previously.\n",
      " |  \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |  \n",
      " |      header -- :class:`~pysam.AlignmentHeader` object to map numerical\n",
      " |                identifiers to chromosome names. If not given, an empty\n",
      " |                header is created.\n",
      " |  \n",
      " |  Methods defined here:\n",
      " |  \n",
      " |  __copy__(...)\n",
      " |      AlignedSegment.__copy__(self)\n",
      " |  \n",
      " |  __deepcopy__(...)\n",
      " |      AlignedSegment.__deepcopy__(self, memo)\n",
      " |  \n",
      " |  __eq__(self, value, /)\n",
      " |      Return self==value.\n",
      " |  \n",
      " |  __ge__(self, value, /)\n",
      " |      Return self>=value.\n",
      " |  \n",
      " |  __gt__(self, value, /)\n",
      " |      Return self>value.\n",
      " |  \n",
      " |  __hash__(self, /)\n",
      " |      Return hash(self).\n",
      " |  \n",
      " |  __init__(self, /, *args, **kwargs)\n",
      " |      Initialize self.  See help(type(self)) for accurate signature.\n",
      " |  \n",
      " |  __le__(self, value, /)\n",
      " |      Return self<=value.\n",
      " |  \n",
      " |  __lt__(self, value, /)\n",
      " |      Return self<value.\n",
      " |  \n",
      " |  __ne__(self, value, /)\n",
      " |      Return self!=value.\n",
      " |  \n",
      " |  __reduce__ = __reduce_cython__(...)\n",
      " |      AlignedSegment.__reduce_cython__(self)\n",
      " |  \n",
      " |  __setstate__ = __setstate_cython__(...)\n",
      " |      AlignedSegment.__setstate_cython__(self, __pyx_state)\n",
      " |  \n",
      " |  __str__(...)\n",
      " |      return string representation of alignment.\n",
      " |      \n",
      " |      The representation is an approximate :term:`SAM` format, because\n",
      " |      an aligned read might not be associated with a :term:`AlignmentFile`.\n",
      " |      As a result :term:`tid` is shown instead of the reference name.\n",
      " |      Similarly, the tags field is returned in its parsed state.\n",
      " |      \n",
      " |      To get a valid SAM record, use :meth:`to_string`.\n",
      " |  \n",
      " |  compare(...)\n",
      " |      AlignedSegment.compare(self, AlignedSegment other)\n",
      " |      return -1,0,1, if contents in this are binary\n",
      " |              <,=,> to *other*\n",
      " |  \n",
      " |  get_aligned_pairs(...)\n",
      " |      AlignedSegment.get_aligned_pairs(self, matches_only=False, with_seq=False)\n",
      " |      a list of aligned read (query) and reference positions.\n",
      " |      \n",
      " |              For inserts, deletions, skipping either query or reference\n",
      " |              position may be None.\n",
      " |      \n",
      " |              Padding is currently not supported and leads to an exception.\n",
      " |      \n",
      " |              Parameters\n",
      " |              ----------\n",
      " |      \n",
      " |              matches_only : bool\n",
      " |                If True, only matched bases are returned - no None on either\n",
      " |                side.\n",
      " |              with_seq : bool\n",
      " |                If True, return a third element in the tuple containing the\n",
      " |                reference sequence. Substitutions are lower-case. This option\n",
      " |                requires an MD tag to be present.\n",
      " |      \n",
      " |              Returns\n",
      " |              -------\n",
      " |      \n",
      " |              aligned_pairs : list of tuples\n",
      " |  \n",
      " |  get_blocks(...)\n",
      " |      AlignedSegment.get_blocks(self)\n",
      " |      a list of start and end positions of\n",
      " |             aligned gapless blocks.\n",
      " |      \n",
      " |             The start and end positions are in genomic\n",
      " |             coordinates.\n",
      " |      \n",
      " |             Blocks are not normalized, i.e. two blocks\n",
      " |             might be directly adjacent. This happens if\n",
      " |             the two blocks are separated by an insertion\n",
      " |             in the read.\n",
      " |  \n",
      " |  get_cigar_stats(...)\n",
      " |      AlignedSegment.get_cigar_stats(self)\n",
      " |      summary of operations in cigar string.\n",
      " |      \n",
      " |              The output order in the array is \"MIDNSHP=X\" followed by a\n",
      " |              field for the NM tag. If the NM tag is not present, this\n",
      " |              field will always be 0.\n",
      " |      \n",
      " |              +-----+--------------+-----+\n",
      " |              |M    |BAM_CMATCH    |0    |\n",
      " |              +-----+--------------+-----+\n",
      " |              |I    |BAM_CINS      |1    |\n",
      " |              +-----+--------------+-----+\n",
      " |              |D    |BAM_CDEL      |2    |\n",
      " |              +-----+--------------+-----+\n",
      " |              |N    |BAM_CREF_SKIP |3    |\n",
      " |              +-----+--------------+-----+\n",
      " |              |S    |BAM_CSOFT_CLIP|4    |\n",
      " |              +-----+--------------+-----+\n",
      " |              |H    |BAM_CHARD_CLIP|5    |\n",
      " |              +-----+--------------+-----+\n",
      " |              |P    |BAM_CPAD      |6    |\n",
      " |              +-----+--------------+-----+\n",
      " |              |=    |BAM_CEQUAL    |7    |\n",
      " |              +-----+--------------+-----+\n",
      " |              |X    |BAM_CDIFF     |8    |\n",
      " |              +-----+--------------+-----+\n",
      " |              |B    |BAM_CBACK     |9    |\n",
      " |              +-----+--------------+-----+\n",
      " |              |NM   |NM tag        |10   |\n",
      " |              +-----+--------------+-----+\n",
      " |      \n",
      " |              If no cigar string is present, empty arrays will be returned.\n",
      " |      \n",
      " |              Parameters\n",
      " |              ----------\n",
      " |      \n",
      " |              Returns\n",
      " |              -------\n",
      " |      \n",
      " |              arrays : two arrays. The first contains the nucleotide counts within\n",
      " |                 each cigar operation, the second contains the number of blocks for\n",
      " |                 each cigar operation.\n",
      " |  \n",
      " |  get_forward_qualities(...)\n",
      " |      AlignedSegment.get_forward_qualities(self)\n",
      " |      return base qualities of the read sequence.\n",
      " |              \n",
      " |              Reads mapping to the reverse strand will be reversed.\n",
      " |  \n",
      " |  get_forward_sequence(...)\n",
      " |      AlignedSegment.get_forward_sequence(self)\n",
      " |      return the original read sequence.\n",
      " |              \n",
      " |              Reads mapping to the reverse strand will be reverse\n",
      " |              complemented.\n",
      " |      \n",
      " |              Returns None if the record has no query sequence.\n",
      " |  \n",
      " |  get_overlap(...)\n",
      " |      AlignedSegment.get_overlap(self, uint32_t start, uint32_t end)\n",
      " |      return number of aligned bases of read overlapping the interval\n",
      " |              *start* and *end* on the reference sequence.\n",
      " |      \n",
      " |              Return None if cigar alignment is not available.\n",
      " |  \n",
      " |  get_reference_positions(...)\n",
      " |      AlignedSegment.get_reference_positions(self, full_length=False)\n",
      " |      a list of reference positions that this read aligns to.\n",
      " |      \n",
      " |              By default, this method only returns positions in the\n",
      " |              reference that are within the alignment. If *full_length* is\n",
      " |              set, None values will be included for any soft-clipped or\n",
      " |              unaligned positions within the read. The returned list will\n",
      " |              thus be of the same length as the read.\n",
      " |  \n",
      " |  get_reference_sequence(...)\n",
      " |      AlignedSegment.get_reference_sequence(self)\n",
      " |      return the reference sequence in the region that is covered by the\n",
      " |              alignment of the read to the reference.\n",
      " |      \n",
      " |              This method requires the MD tag to be set.\n",
      " |  \n",
      " |  get_tag(...)\n",
      " |      AlignedSegment.get_tag(self, tag, with_value_type=False)\n",
      " |      \n",
      " |      retrieves data from the optional alignment section\n",
      " |      given a two-letter *tag* denoting the field.\n",
      " |      \n",
      " |      The returned value is cast into an appropriate python type.\n",
      " |      \n",
      " |      This method is the fastest way to access the optional\n",
      " |      alignment section if only few tags need to be retrieved.\n",
      " |      \n",
      " |      Possible value types are \"AcCsSiIfZHB\" (see BAM format\n",
      " |      specification) as well as additional value type 'd' as\n",
      " |      implemented in htslib.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      \n",
      " |      tag :\n",
      " |          data tag.\n",
      " |      \n",
      " |      with_value_type : Optional[bool]\n",
      " |          if set to True, the return value is a tuple of (tag value, type code).\n",
      " |          (default False)\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      \n",
      " |      A python object with the value of the `tag`. The type of the\n",
      " |      object depends on the data type in the data record.\n",
      " |      \n",
      " |      Raises\n",
      " |      ------\n",
      " |      \n",
      " |      KeyError\n",
      " |          If `tag` is not present, a KeyError is raised.\n",
      " |  \n",
      " |  get_tags(...)\n",
      " |      AlignedSegment.get_tags(self, with_value_type=False)\n",
      " |      the fields in the optional aligment section.\n",
      " |      \n",
      " |              Returns a list of all fields in the optional\n",
      " |              alignment section. Values are converted to appropriate python\n",
      " |              values. For example:\n",
      " |      \n",
      " |              [(NM, 2), (RG, \"GJP00TM04\")]\n",
      " |      \n",
      " |              If *with_value_type* is set, the value type as encode in\n",
      " |              the AlignedSegment record will be returned as well:\n",
      " |      \n",
      " |              [(NM, 2, \"i\"), (RG, \"GJP00TM04\", \"Z\")]\n",
      " |      \n",
      " |              This method will convert all values in the optional alignment\n",
      " |              section. When getting only one or few tags, please see\n",
      " |              :meth:`get_tag` for a quicker way to achieve this.\n",
      " |  \n",
      " |  has_tag(...)\n",
      " |      AlignedSegment.has_tag(self, tag)\n",
      " |      returns true if the optional alignment section\n",
      " |              contains a given *tag*.\n",
      " |  \n",
      " |  infer_query_length(...)\n",
      " |      AlignedSegment.infer_query_length(self, always=False)\n",
      " |      infer query length from CIGAR alignment.\n",
      " |      \n",
      " |              This method deduces the query length from the CIGAR alignment\n",
      " |              but does not include hard-clipped bases.\n",
      " |      \n",
      " |              Returns None if CIGAR alignment is not present.\n",
      " |      \n",
      " |              If *always* is set to True, `infer_read_length` is used instead.\n",
      " |              This is deprecated and only present for backward compatibility.\n",
      " |  \n",
      " |  infer_read_length(...)\n",
      " |      AlignedSegment.infer_read_length(self)\n",
      " |      infer read length from CIGAR alignment.\n",
      " |      \n",
      " |              This method deduces the read length from the CIGAR alignment\n",
      " |              including hard-clipped bases.\n",
      " |      \n",
      " |              Returns None if CIGAR alignment is not present.\n",
      " |  \n",
      " |  opt(...)\n",
      " |      AlignedSegment.opt(self, tag)\n",
      " |      deprecated, use get_tag() instead\n",
      " |  \n",
      " |  overlap(...)\n",
      " |      AlignedSegment.overlap(self)\n",
      " |      deprecated, use get_overlap() instead\n",
      " |  \n",
      " |  setTag(...)\n",
      " |      AlignedSegment.setTag(self, tag, value, value_type=None, replace=True)\n",
      " |      deprecated, use set_tag() instead\n",
      " |  \n",
      " |  set_tag(...)\n",
      " |      AlignedSegment.set_tag(self, tag, value, value_type=None, replace=True)\n",
      " |      sets a particular field *tag* to *value* in the optional alignment\n",
      " |              section.\n",
      " |      \n",
      " |              *value_type* describes the type of *value* that is to entered\n",
      " |              into the alignment record. It can be set explicitly to one of\n",
      " |              the valid one-letter type codes. If unset, an appropriate type\n",
      " |              will be chosen automatically based on the python type of\n",
      " |              *value*.\n",
      " |      \n",
      " |              An existing value of the same *tag* will be overwritten unless\n",
      " |              *replace* is set to False. This is usually not recommened as a\n",
      " |              tag may only appear once in the optional alignment section.\n",
      " |      \n",
      " |              If *value* is None, the tag will be deleted.\n",
      " |      \n",
      " |              This method accepts valid SAM specification value types, which\n",
      " |              are::\n",
      " |              \n",
      " |                 A: printable char\n",
      " |                 i: signed int\n",
      " |                 f: float\n",
      " |                 Z: printable string\n",
      " |                 H: Byte array in hex format\n",
      " |                 B: Integer or numeric array\n",
      " |      \n",
      " |              Additionally, it will accept the integer BAM types ('cCsSI')\n",
      " |      \n",
      " |              For htslib compatibility, 'a' is synonymous with 'A' and the\n",
      " |              method accepts a 'd' type code for a double precision float.\n",
      " |      \n",
      " |              When deducing the type code by the python type of *value*, the\n",
      " |              following mapping is applied::\n",
      " |              \n",
      " |                  i: python int\n",
      " |                  f: python float\n",
      " |                  Z: python str or bytes\n",
      " |                  B: python array.array, list or tuple\n",
      " |                  \n",
      " |              Note that a single character string will be output as 'Z' and\n",
      " |              not 'A' as the former is the more general type.\n",
      " |  \n",
      " |  set_tags(...)\n",
      " |      AlignedSegment.set_tags(self, tags)\n",
      " |      sets the fields in the optional alignment section with\n",
      " |              a list of (tag, value) tuples.\n",
      " |      \n",
      " |              The :term:`value type` of the values is determined from the\n",
      " |              python type. Optionally, a type may be given explicitly as\n",
      " |              a third value in the tuple, For example:\n",
      " |      \n",
      " |              x.set_tags([(NM, 2, \"i\"), (RG, \"GJP00TM04\", \"Z\")]\n",
      " |      \n",
      " |              This method will not enforce the rule that the same tag may appear\n",
      " |              only once in the optional alignment section.\n",
      " |  \n",
      " |  to_dict(...)\n",
      " |      AlignedSegment.to_dict(self)\n",
      " |      returns a json representation of the aligned segment.\n",
      " |      \n",
      " |              Field names are abbreviated versions of the class attributes.\n",
      " |  \n",
      " |  to_string(...)\n",
      " |      AlignedSegment.to_string(self)\n",
      " |      returns a string representation of the aligned segment.\n",
      " |      \n",
      " |              The output format is valid SAM format if a header is associated\n",
      " |              with the AlignedSegment.\n",
      " |  \n",
      " |  tostring(...)\n",
      " |      AlignedSegment.tostring(self, htsfile=None)\n",
      " |      deprecated, use :meth:`to_string()` instead.\n",
      " |      \n",
      " |              Parameters\n",
      " |              ----------\n",
      " |      \n",
      " |              htsfile -- (deprecated) AlignmentFile object to map numerical\n",
      " |                         identifiers to chromosome names. This parameter is present\n",
      " |                         for backwards compatibility and ignored.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Class methods defined here:\n",
      " |  \n",
      " |  from_dict(...) from builtins.type\n",
      " |      AlignedSegment.from_dict(type cls, sam_dict, AlignmentHeader header)\n",
      " |      parses a dictionary representation of the aligned segment.\n",
      " |      \n",
      " |              Parameters\n",
      " |              ----------\n",
      " |              sam_dict -- dictionary of alignment values, keys corresponding to output from\n",
      " |                          :meth:`todict()`.\n",
      " |  \n",
      " |  fromstring(...) from builtins.type\n",
      " |      AlignedSegment.fromstring(type cls, sam, AlignmentHeader header)\n",
      " |      parses a string representation of the aligned segment.\n",
      " |      \n",
      " |              The input format should be valid SAM format.\n",
      " |      \n",
      " |              Parameters\n",
      " |              ----------\n",
      " |              sam -- :term:`SAM` formatted string\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Static methods defined here:\n",
      " |  \n",
      " |  __new__(*args, **kwargs) from builtins.type\n",
      " |      Create and return a new object.  See help(type) for accurate signature.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors defined here:\n",
      " |  \n",
      " |  aend\n",
      " |      deprecated, reference_end instead\n",
      " |  \n",
      " |  alen\n",
      " |      deprecated, reference_length instead\n",
      " |  \n",
      " |  aligned_pairs\n",
      " |      deprecated, use get_aligned_pairs() instead\n",
      " |  \n",
      " |  bin\n",
      " |      properties bin\n",
      " |  \n",
      " |  blocks\n",
      " |      deprecated, use get_blocks() instead\n",
      " |  \n",
      " |  cigar\n",
      " |      deprecated, use cigartuples instead\n",
      " |  \n",
      " |  cigarstring\n",
      " |      the :term:`cigar` alignment as a string.\n",
      " |      \n",
      " |      The cigar string is a string of alternating integers\n",
      " |      and characters denoting the length and the type of\n",
      " |      an operation.\n",
      " |      \n",
      " |      .. note::\n",
      " |          The order length,operation is specified in the\n",
      " |          SAM format. It is different from the order of\n",
      " |          the :attr:`cigar` property.\n",
      " |      \n",
      " |      Returns None if not present.\n",
      " |      \n",
      " |      To unset the cigarstring, assign None or the\n",
      " |      empty string.\n",
      " |  \n",
      " |  cigartuples\n",
      " |      the :term:`cigar` alignment. The alignment\n",
      " |      is returned as a list of tuples of (operation, length).\n",
      " |      \n",
      " |      If the alignment is not present, None is returned.\n",
      " |      \n",
      " |      The operations are:\n",
      " |      \n",
      " |      +-----+--------------+-----+\n",
      " |      |M    |BAM_CMATCH    |0    |\n",
      " |      +-----+--------------+-----+\n",
      " |      |I    |BAM_CINS      |1    |\n",
      " |      +-----+--------------+-----+\n",
      " |      |D    |BAM_CDEL      |2    |\n",
      " |      +-----+--------------+-----+\n",
      " |      |N    |BAM_CREF_SKIP |3    |\n",
      " |      +-----+--------------+-----+\n",
      " |      |S    |BAM_CSOFT_CLIP|4    |\n",
      " |      +-----+--------------+-----+\n",
      " |      |H    |BAM_CHARD_CLIP|5    |\n",
      " |      +-----+--------------+-----+\n",
      " |      |P    |BAM_CPAD      |6    |\n",
      " |      +-----+--------------+-----+\n",
      " |      |=    |BAM_CEQUAL    |7    |\n",
      " |      +-----+--------------+-----+\n",
      " |      |X    |BAM_CDIFF     |8    |\n",
      " |      +-----+--------------+-----+\n",
      " |      |B    |BAM_CBACK     |9    |\n",
      " |      +-----+--------------+-----+\n",
      " |      \n",
      " |      .. note::\n",
      " |          The output is a list of (operation, length) tuples, such as\n",
      " |          ``[(0, 30)]``.\n",
      " |          This is different from the SAM specification and\n",
      " |          the :attr:`cigarstring` property, which uses a\n",
      " |          (length, operation) order, for example: ``30M``.\n",
      " |      \n",
      " |      To unset the cigar property, assign an empty list\n",
      " |      or None.\n",
      " |  \n",
      " |  flag\n",
      " |      properties flag\n",
      " |  \n",
      " |  header\n",
      " |  \n",
      " |  inferred_length\n",
      " |      deprecated, use infer_query_length() instead\n",
      " |  \n",
      " |  is_duplicate\n",
      " |      true if optical or PCR duplicate\n",
      " |  \n",
      " |  is_paired\n",
      " |      true if read is paired in sequencing\n",
      " |  \n",
      " |  is_proper_pair\n",
      " |      true if read is mapped in a proper pair\n",
      " |  \n",
      " |  is_qcfail\n",
      " |      true if QC failure\n",
      " |  \n",
      " |  is_read1\n",
      " |      true if this is read1\n",
      " |  \n",
      " |  is_read2\n",
      " |      true if this is read2\n",
      " |  \n",
      " |  is_reverse\n",
      " |      true if read is mapped to reverse strand\n",
      " |  \n",
      " |  is_secondary\n",
      " |      true if not primary alignment\n",
      " |  \n",
      " |  is_supplementary\n",
      " |      true if this is a supplementary alignment\n",
      " |  \n",
      " |  is_unmapped\n",
      " |      true if read itself is unmapped\n",
      " |  \n",
      " |  isize\n",
      " |      deprecated, use template_length instead\n",
      " |  \n",
      " |  mapping_quality\n",
      " |      mapping quality\n",
      " |  \n",
      " |  mapq\n",
      " |      deprecated, use mapping_quality instead\n",
      " |  \n",
      " |  mate_is_reverse\n",
      " |      true is read is mapped to reverse strand\n",
      " |  \n",
      " |  mate_is_unmapped\n",
      " |      true if the mate is unmapped\n",
      " |  \n",
      " |  mpos\n",
      " |      deprecated, use next_reference_start instead\n",
      " |  \n",
      " |  mrnm\n",
      " |      deprecated, use next_reference_id instead\n",
      " |  \n",
      " |  next_reference_id\n",
      " |      the :term:`reference` id of the mate/next read.\n",
      " |  \n",
      " |  next_reference_name\n",
      " |      :term:`reference` name of the mate/next read (None if no\n",
      " |      AlignmentFile is associated)\n",
      " |  \n",
      " |  next_reference_start\n",
      " |      the position of the mate/next read.\n",
      " |  \n",
      " |  pnext\n",
      " |      deprecated, use next_reference_start instead\n",
      " |  \n",
      " |  pos\n",
      " |      deprecated, use reference_start instead\n",
      " |  \n",
      " |  positions\n",
      " |      deprecated, use get_reference_positions() instead\n",
      " |  \n",
      " |  qend\n",
      " |      deprecated, use query_alignment_end instead\n",
      " |  \n",
      " |  qlen\n",
      " |      deprecated, use query_alignment_length instead\n",
      " |  \n",
      " |  qname\n",
      " |      deprecated, use query_name instead\n",
      " |  \n",
      " |  qqual\n",
      " |      deprecated, query_alignment_qualities instead\n",
      " |  \n",
      " |  qstart\n",
      " |      deprecated, use query_alignment_start instead\n",
      " |  \n",
      " |  qual\n",
      " |      deprecated, query_qualities instead\n",
      " |  \n",
      " |  query\n",
      " |      deprecated, query_alignment_sequence instead\n",
      " |  \n",
      " |  query_alignment_end\n",
      " |      end index of the aligned query portion of the sequence (0-based,\n",
      " |      exclusive)\n",
      " |      \n",
      " |      This the index just past the last base in :attr:`seq` that is not\n",
      " |      soft-clipped.\n",
      " |  \n",
      " |  query_alignment_length\n",
      " |      length of the aligned query sequence.\n",
      " |      \n",
      " |      This is equal to :attr:`qend` - :attr:`qstart`\n",
      " |  \n",
      " |  query_alignment_qualities\n",
      " |      aligned query sequence quality values (None if not present). These\n",
      " |      are the quality values that correspond to :attr:`query`, that\n",
      " |      is, they exclude qualities of :term:`soft clipped` bases. This\n",
      " |      is equal to ``qual[qstart:qend]``.\n",
      " |      \n",
      " |      Quality scores are returned as a python array of unsigned\n",
      " |      chars. Note that this is not the ASCII-encoded value typically\n",
      " |      seen in FASTQ or SAM formatted files. Thus, no offset of 33\n",
      " |      needs to be subtracted.\n",
      " |      \n",
      " |      This property is read-only.\n",
      " |  \n",
      " |  query_alignment_sequence\n",
      " |      aligned portion of the read.\n",
      " |      \n",
      " |      This is a substring of :attr:`seq` that excludes flanking\n",
      " |      bases that were :term:`soft clipped` (None if not present). It\n",
      " |      is equal to ``seq[qstart:qend]``.\n",
      " |      \n",
      " |      SAM/BAM files may include extra flanking bases that are not\n",
      " |      part of the alignment.  These bases may be the result of the\n",
      " |      Smith-Waterman or other algorithms, which may not require\n",
      " |      alignments that begin at the first residue or end at the last.\n",
      " |      In addition, extra sequencing adapters, multiplex identifiers,\n",
      " |      and low-quality bases that were not considered for alignment\n",
      " |      may have been retained.\n",
      " |  \n",
      " |  query_alignment_start\n",
      " |      start index of the aligned query portion of the sequence (0-based,\n",
      " |      inclusive).\n",
      " |      \n",
      " |      This the index of the first base in :attr:`seq` that is not\n",
      " |      soft-clipped.\n",
      " |  \n",
      " |  query_length\n",
      " |      the length of the query/read.\n",
      " |      \n",
      " |      This value corresponds to the length of the sequence supplied\n",
      " |      in the BAM/SAM file. The length of a query is 0 if there is no\n",
      " |      sequence in the BAM/SAM file. In those cases, the read length\n",
      " |      can be inferred from the CIGAR alignment, see\n",
      " |      :meth:`pysam.AlignedSegment.infer_query_length`.\n",
      " |      \n",
      " |      The length includes soft-clipped bases and is equal to\n",
      " |      ``len(query_sequence)``.\n",
      " |      \n",
      " |      This property is read-only but can be set by providing a\n",
      " |      sequence.\n",
      " |      \n",
      " |      Returns 0 if not available.\n",
      " |  \n",
      " |  query_name\n",
      " |      the query template name (None if not present)\n",
      " |  \n",
      " |  query_qualities\n",
      " |      read sequence base qualities, including :term:`soft\n",
      " |      clipped` bases (None if not present).\n",
      " |      \n",
      " |      Quality scores are returned as a python array of unsigned\n",
      " |      chars. Note that this is not the ASCII-encoded value typically\n",
      " |      seen in FASTQ or SAM formatted files. Thus, no offset of 33\n",
      " |      needs to be subtracted.\n",
      " |      \n",
      " |      Note that to set quality scores the sequence has to be set\n",
      " |      beforehand as this will determine the expected length of the\n",
      " |      quality score array.\n",
      " |      \n",
      " |      This method raises a ValueError if the length of the\n",
      " |      quality scores and the sequence are not the same.\n",
      " |  \n",
      " |  query_sequence\n",
      " |      read sequence bases, including :term:`soft clipped` bases\n",
      " |      (None if not present).\n",
      " |      \n",
      " |      Note that assigning to seq will invalidate any quality scores.\n",
      " |      Thus, to in-place edit the sequence and quality scores, copies of\n",
      " |      the quality scores need to be taken. Consider trimming for example::\n",
      " |      \n",
      " |         q = read.query_qualities\n",
      " |         read.query_squence = read.query_sequence[5:10]\n",
      " |         read.query_qualities = q[5:10]\n",
      " |      \n",
      " |      The sequence is returned as it is stored in the BAM file. Some mappers\n",
      " |      might have stored a reverse complement of the original read\n",
      " |      sequence.\n",
      " |  \n",
      " |  reference_end\n",
      " |      aligned reference position of the read on the reference genome.\n",
      " |      \n",
      " |      reference_end points to one past the last aligned residue.\n",
      " |      Returns None if not available (read is unmapped or no cigar\n",
      " |      alignment present).\n",
      " |  \n",
      " |  reference_id\n",
      " |      :term:`reference` ID\n",
      " |      \n",
      " |      .. note::\n",
      " |      \n",
      " |          This field contains the index of the reference sequence in\n",
      " |          the sequence dictionary. To obtain the name of the\n",
      " |          reference sequence, use :meth:`get_reference_name()`\n",
      " |  \n",
      " |  reference_length\n",
      " |      aligned length of the read on the reference genome.\n",
      " |      \n",
      " |      This is equal to `aend - pos`. Returns None if not available.\n",
      " |  \n",
      " |  reference_name\n",
      " |      :term:`reference` name\n",
      " |  \n",
      " |  reference_start\n",
      " |      0-based leftmost coordinate\n",
      " |  \n",
      " |  rlen\n",
      " |      deprecated, query_length instead\n",
      " |  \n",
      " |  rname\n",
      " |      deprecated, use reference_id instead\n",
      " |  \n",
      " |  rnext\n",
      " |      deprecated, use next_reference_id instead\n",
      " |  \n",
      " |  seq\n",
      " |      deprecated, use query_sequence instead\n",
      " |  \n",
      " |  tags\n",
      " |      deprecated, use get_tags() instead\n",
      " |  \n",
      " |  template_length\n",
      " |      the observed query template length\n",
      " |  \n",
      " |  tid\n",
      " |      deprecated, use reference_id instead\n",
      " |  \n",
      " |  tlen\n",
      " |      deprecated, use template_length instead\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data and other attributes defined here:\n",
      " |  \n",
      " |  __pyx_vtable__ = <capsule object NULL>\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<pysam.libcalignmentfile.AlignmentHeader at 0x11aa207b0>"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bam_file = \"data/IL27/flye.bam\"\n",
    "bam = pysam.AlignmentFile(bam_file, 'rb')\n",
    "contigs = [contig for contig in bam.fetch()]\n",
    "help(contigs[1])\n",
    "contigs[1].header"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('NM', 1656),\n",
       " ('ms', 460610),\n",
       " ('AS', 460632),\n",
       " ('nn', 0),\n",
       " ('tp', 'P'),\n",
       " ('cm', 37508),\n",
       " ('s1', 210541),\n",
       " ('s2', 788),\n",
       " ('de', 0.0015999999595806003)]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "contig = contigs[1]\n",
    "tags = contig.get_tags()\n",
    "tags"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "list"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "type(tags)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1656"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "contig.get_tag('NM')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
